pytket/extensions/cutensornet/mps/mps.py

```diff
@@ -446,38 +452,69 @@ def vdot(self, other: MPS) -> complex:
-                    "Ll,LRp,lrp->Rr",
-                    partial_result,
-                    self.tensors[pos].conj(),
-                    other.tensors[pos],
-                )
-            # Finally, contract the last tensor
-            result = cq.contract(
-                "Ll,LRp,lrp->",  # R and r are dim 1, so they are omitted; scalar result
-                partial_result,
-                self.tensors[-1].conj(),
-                other.tensors[-1],
-            )
+        self._logger.debug("Applying vdot between two MPS.")
+
+        # We convert both MPS to their interleaved representation and
+        # contract them using cuQuantum.
+        mps1 = self._get_interleaved_representation(conj=True)
+        mps2 = other._get_interleaved_representation(conj=False)
+        interleaved_rep = mps1 + mps2
+        interleaved_rep.append([])  # Discards dim=1 bonds with []
+
+        # We define the contraction path ourselves
+        end_mps1 = len(self) - 1  # Rightmost tensor of mps1 in interleaved_rep
+        end_mps2 = len(self) + len(other) - 1  # Rightmost tensor of mps2
+        contraction_path = [(end_mps1, end_mps2)]  # Contract ends of mps1 and mps2
+        for _ in range(len(self) - 1):
+            # Update the position markers
+            end_mps1 -= 1  # One tensor was removed from mps1
+            end_mps2 -= 2  # One tensor removed from mps1 and another from mps2
+            # Contract the result from last iteration with the ends of mps1 and mps2
+            contraction_path.append((end_mps2, end_mps2 + 1))  # End of mps2 and result
+            contraction_path.append((end_mps1, end_mps2))  # End of mps1 and ^ outcome
+
+        # Apply the contraction
+        result = cq.contract(
+            *interleaved_rep,
+            options={"handle": self._lib.handle, "device_id": self._lib.device_id},
+            optimize={"path": contraction_path},
+        )
 
         self._logger.debug(f"Result from vdot={result}")
         return complex(result)
 
+    def _get_interleaved_representation(
+        self, conj: bool = False
+    ) -> list[Union[cp.ndarray, str]]:
+        """Returns the interleaved representation of the MPS used by cuQuantum.
+
+        Args:
+            conj: If True, all tensors are conjugated and bonds IDs are prefixed
+                with * (except physical bonds). Defaults to False.
+        """
+        self._logger.debug("Creating interleaved representation...")
+
+        # Auxiliar dictionary of physical bonds to qubit IDs
+        qubit_id = {location: qubit for qubit, location in self.qubit_position.items()}
+
+        interleaved_rep = []
+        for i, t in enumerate(self.tensors):
+            # Append the tensor
+            if conj:
+                interleaved_rep.append(t.conj())
+            else:
+                interleaved_rep.append(t)
+
+            # Create the ID for the bonds involved
+            bonds = [str(i), str(i + 1), str(qubit_id[i])]

```

Comment:
Sorry for a stupid question, but why are there three elements? Are there three indices to contract along?

Reply:
These are the bonds of a single MPS tensor. All MPS tensors have a left bond, a right bond and a physical bond, those are the IDs in that order. The leftmost (and rightmost) tensor also has three bonds, it's just that the left one (right one, respectively) has dimension 1, so it's a dummy bond.

Reply:
But of course, thank you!